x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -43,71 +42,84 @@
-    private final Integer cacheMaxUsers;
+    /*package private*/ final TimeValue cacheTtl;
+    /*package private*/ final Integer cacheMaxUsers;
     // Standard HTTP settings for outgoing connections to get JWT issuer jwkset_path
-    private final TimeValue httpConnectTimeout;
-    private final TimeValue httpConnectionReadTimeout;
-    private final TimeValue httpSocketTimeout;
-    private final Integer httpMaxConnections;
-    private final Integer httpMaxEndpointConnections;
-
-    // constructor derives members
-    private final URL jwkSetPathUrl; // Non-null if jwkSetPath is set and starts with "https://"
-    private final Path jwkSetPathObj; // Non-null if jwkSetPath is set and resolves to a local config file
-    private final Cache<String, ListenableFuture<CachedAuthenticationSuccess>> cachedAuthenticationSuccesses;
-    private final Hasher hasher;
+    /*package private*/ final TimeValue httpConnectTimeout;
+    /*package private*/ final TimeValue httpConnectionReadTimeout;
+    /*package private*/ final TimeValue httpSocketTimeout;
+    /*package private*/ final Integer httpMaxConnections;
+    /*package private*/ final Integer httpMaxEndpointConnections;
+    // Standard HTTP proxy settings for outgoing connections to get JWT issuer jwkset_path
+    /*package private*/ final String httpProxyScheme;
+    /*package private*/ final int httpProxyPort;
+    /*package private*/ final String httpProxyHost;
+
+    // Helpers: Constructor creates these members
+    /*package private*/ final Hasher hasher;
+    /*package private*/ final Cache<String, CacheResult> cache;
+    /*package private*/ final CloseableHttpAsyncClient httpClient;
+    /*package private*/ final JWKSet jwkSetHmac;
+    /*package private*/ final JWKSet jwkSetPkc;
 
     // initialize sets this value, not the constructor, because all realms objects need to be constructed before linking any delegates
-    private DelegatedAuthorizationSupport delegatedAuthorizationSupport;
+    /*package private*/ DelegatedAuthorizationSupport delegatedAuthorizationSupport;
+
+    // usage stats
+    /*package private*/ final AtomicLong counterTokenFail = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterTokenSuccess = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterAuthenticateFail = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterAuthenticateSuccess = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterCacheGetFail = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterCacheGetOkWarning = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterCacheGetOkNoWarning = new AtomicLong(0);
+    /*package private*/ final AtomicLong counterCacheAdd = new AtomicLong(0);
+
+    // Only JwtRealmTest.addJwtRealm() uses these package private members, to generate test JWTs from these test JWKs and Users.
+    /*package private*/ List<JWK> testIssuerJwks = null;
```

Comment:
For JWT testing, a test issuer needs to know at least two settings not available to JWT realms.
1. All private/public/secret JWKs. Realm only has all **public/secret JWKs.
2. All usernames and roles, to know what random generated roles to except from role mapping or authz lookup.

The code for using the lists of JWKs and users/roles starts around JwtRealmTests.java line 190.

Generating a JWT Realm is handled in addRealm(). To return the issuer settings, I considered these two choices:
- Return the JwtRealm, and attach the issuer test settings as package private, or
- Return a Tuple<JwtRealm,Issuer>, so the issuer settings are in a separate class

I didn't know which way PR review would go, so I took the easier option and left this as a discussion point. If you think moving those two things to a separate class are worthwhile, I can do it.

Reply:
I stand by my opinion that we should not be putting test related code in the realm class. I think this can be significantly simplified :

- In `final List<JWK> testJwks = JwtTestCase.toRandomJwks(JwtUtil.toJwsAlgorithms(signatureAlgorithmList));` you generate `testJwks` with variables that you are passing  in `addJwtRealm()` ( namely `signatureAlgorithmList` ) 
- in 
  ````
  final Map<String, User> testUsers = JwtTestCase.generateTestUsersWithRoles(
            randomIntBetween(5, 10),
            randomIntBetween(rolesRange.v1(), rolesRange.v2())
        );
  ```     
  you generate `testUsers` with variables that you are passing in `addJwtRealm()` ( namely `rolesRange`) .
  
Can't we generate testJwks and testUsers, _before_ calling `addJwtRealm` instead of passing the values to `addJwtRealm` to generate these for us ? Might even be simpler/different ways to do this, I have no strong opinions, as long as we remove the test variables from production code

Reply:
I will remove the test variables from the realm. It will take a bit of refactoring but I agree it will improve the code.